src/middlewares/createInsightsMiddleware.ts

```diff
@@ -88,25 +90,29 @@ export const createInsightsMiddleware: CreateInsightsMiddleware = (props) => {
     });
     insightsClient('init', { appId, apiKey, ...insightsInitParams });
 
+    const createWidget = connectConfigure(noop);
+    let widget: ReturnType<typeof createWidget>;
+
     return {
       onStateChange() {},
       subscribe() {
         insightsClient('addAlgoliaAgent', 'insights-middleware');
 
-        // At the time this middleware is subscribed, `mainIndex.init()` is already called.
-        // It means `mainIndex.getHelper()` exists.
-        const helper = instantSearchInstance.mainIndex.getHelper()!;
+        let searchParameters: Partial<PlainSearchParameters> = {
+          clickAnalytics: true,
+        };
+        widget = createWidget({ searchParameters });
```

Comment:
I'd create the arguments dynamically, but not the widget twice, that way you can add the widget with the known arguments *after* the if (or is that not possible?)

Reply:
```diff
        let searchParameters: Partial<PlainSearchParameters> = {
          clickAnalytics: true,
        };
-       widget = createWidget({ searchParameters });
-       instantSearchInstance.addWidgets([widget]);

        const setUserTokenToSearch = (userToken?: string) => {
          instantSearchInstance.removeWidgets([widget]);
          searchParameters = {
            ...searchParameters,
            userToken,
          };
          widget = createWidget({ searchParameters });
          instantSearchInstance.addWidgets([widget]);
        };

        const anonymousUserToken = getInsightsAnonymousUserTokenInternal();
        if (hasInsightsClient && anonymousUserToken) {
          // When `aa('init', { ... })` is called, it creates an anonymous user token in cookie.
          // We can set it as userToken.
          setUserTokenToSearch(anonymousUserToken);
        } else {
+         widget = createWidget({ searchParameters });
+         instantSearchInstance.addWidgets([widget]);
        }
```

something like this?

Reply:
I guess I'm not really sure what the flow should be. In that diff, aren't you adding the widget twice?

Reply:
Yep, the widget is added multiple times now. Instead maybe create two widgets, one for userToken and one for clickAnalytics?

Reply:
In the diff or not, it's not adding the widget twice, unless I misunderstand something. Inside `setUserTokenToSearch` function, it removes the existing widget and then accumulate searchParameters with new userToken and add the new widget again.

Reply:
🤔 I missed the removing. It still seems much more complicated than needed, could it not just be two widgets, as configure already merges?

Reply:
Okay, now we're on the similar page. In which part do you feel complicated? That conditional part? Or adding/removing widget?

If the former, indeed I can put two widgets, one for clickAnalytics and another one for userToken. I just want to make sure, we both are aware that although we do that, we still need to remove/add the widget for userToken when we want to update it.

Reply:
adding and removing the widget seems much more complicated than the previous code which used the helper. I feel like the widget's refine could make sense to use (make sure to check it doesn't erase parameters of another configure widget)

Reply:
I almost feel as if there should be a lifecycle event *before* the index init happened (.use essentially) for routing, but one after init has happened to use the helper

Reply:
Yeah, you're right. I did try using `refine` function from the widget, but that's given after `render()` function of the widget is called, which is, again, too late because instantSearchInstance is already started and a API call is fired.

I thought of having additional lifecycle method like `beforeInit` so that the router middleware can manipulate initialUIState. But I had a hard time thinking of another good use case to justify the addition of the API.

What do you think? Should we hold this and have a short meeting when François comes back?